lib/IRGen/IRGenDebugInfo.cpp

```diff
@@ -2338,6 +2330,106 @@ createSpecializedStructOrClassType(NominalOrBoundGenericNominalType *Type,
     }
   }
 
+  /// Returns true if the type's mangled name is affected by an
+  /// @_originallyDefinedIn annotation. This annotation can be on the type
+  /// itself, one of its generic arguments, etc.
+  bool containsOriginallyDefinedIn(Type T) {
+    if (auto *MT = llvm::dyn_cast<MetatypeType>(T))
+      if (containsOriginallyDefinedIn(MT->getInstanceType()))
+        return true;
+    ;
+
+    auto *TypeDecl = T->getNominalOrBoundGenericNominal();
+    if (!TypeDecl)
+      return false;
+
+    // Find the outermost type, since only those can be annotated with
+    // @_originallyDefinedIn.
+    NominalTypeDecl *ParentDecl = TypeDecl;
+    while (llvm::isa_and_nonnull<NominalTypeDecl>(ParentDecl->getParent()))
+      ParentDecl = llvm::cast<NominalTypeDecl>(ParentDecl->getParent());
+
+    if (ParentDecl->getAttrs().hasAttribute<OriginallyDefinedInAttr>())
+      return true;
+
+    // If the type is a bound generic, the type of the substituted generic
+    // arguments might be annotated with @_originallyDefinedIn.
+    if (auto *BGT = llvm::dyn_cast<BoundGenericType>(T))
```

Comment:
I would have expected this function to be implemented in terms of a type visitor. For example, does this catch a a tuple containing an originallyDefinedIn type?